src/client/app/actions/readings.js

```diff
@@ -6,58 +6,125 @@
+	if (readingsForMeterID === undefined) {
+		return true;
+	}
+	if (readingsForMeterID[timeInterval] === undefined) {
+		return true;
+	}
+	if (readingsForMeterID[timeInterval][barDuration] === undefined) {
+		return true;
+	}
+	const readingsForTimeIntervalAndDuration = state.readings.bar.byMeterID[meterID][timeInterval][barDuration];
+	return readingsForTimeIntervalAndDuration === undefined && !readingsForTimeIntervalAndDuration.isFetching;
 }
 
-function receiveManyReadings(meterIDs, timeInterval, readings) {
-	return { type: RECEIVE_MANY_READINGS, meterIDs, timeInterval, readings };
+function requestLineReadings(meterIDs, timeInterval) {
+	return { type: REQUEST_LINE_READINGS, meterIDs, timeInterval };
 }
 
-function fetchManyReadings(meterIDs, timeInterval) {
+function receiveLineReadings(meterIDs, timeInterval, readings) {
+	return { type: RECEIVE_LINE_READINGS, meterIDs, timeInterval, readings };
+}
+
+function requestBarReadings(meterIDs, timeInterval, barDuration) {
+	return { type: REQUEST_BAR_READINGS, meterIDs, timeInterval, barDuration };
+}
+
+function receiveBarReadings(meterIDs, timeInterval, barDuration, readings) {
+	return { type: RECEIVE_BAR_READINGS, meterIDs, timeInterval, barDuration, readings };
+}
+
+function fetchLineReadings(meterIDs, timeInterval) {
 	return dispatch => {
-		dispatch(requestManyReadings(meterIDs, timeInterval));
+		dispatch(requestLineReadings(meterIDs, timeInterval));
 		// The api expects the meter ids to be a comma-separated list.
 		const stringifiedMeterIDs = meterIDs.join(',');
-		return axios.get(`/api/meters/readings/${stringifiedMeterIDs}`, {
+		return axios.get(`/api/readings/line/${stringifiedMeterIDs}`, {
 			params: { timeInterval: timeInterval.toString() }
-		}).then(response => dispatch(receiveManyReadings(meterIDs, timeInterval, response.data)));
+		}).then(response => dispatch(receiveLineReadings(meterIDs, timeInterval, response.data)));
+	};
+}
+
+function fetchBarReadings(meterIDs, timeInterval) {
+	return (dispatch, getState) => {
+		const barDuration = getState().graph.barDuration;
+		dispatch(requestBarReadings(meterIDs, timeInterval, barDuration));
+		const stringifiedMeterIDs = meterIDs.join(',');
+		return axios.get(`/api/readings/bar/${stringifiedMeterIDs}`, {
+			params: { timeInterval: timeInterval.toString(), barDuration }
+		}).then(response => dispatch(receiveBarReadings(meterIDs, timeInterval, barDuration, response.data)));
 	};
 }
 
 /**
- * Fetches readings for the given meterIDs if they are not already fetched or being fetched
+ * Fetches readings for all charts for the given meterIDs if they are not already fetched or being fetched
  * @param {Array.<int>} meterIDs
- * @param {TimeInterval} timeInterval The time interval to fetch readings for
+ * @param {TimeInterval} timeInterval The time interval to fetch readings for on the line chart
+ * @return {*} An action to fetch the needed readings
+ */
+export function fetchAllNeededReadings(meterIDs, timeInterval) {
```

Comment:
This seems to duplicate some functionality from the `fetchBarNeededReadings` function. Maybe extract the line readings fetch logic into a `fetchLineNeededReadings` and make this method just call both?

Also I think that `fetch[Line|Bar]NeededReadings` sounds worse than `fetchNeeded[Line|Bar]Readings`. Should we switch them?